core/src/either.rs

```diff
@@ -493,26 +452,37 @@ where
 {
     type Output = EitherOutput<A::Output, B::Output>;
     type Error = EitherError<A::Error, B::Error>;
-    type Listener = EitherListenStream<A::Listener, B::Listener>;
     type ListenerUpgrade = EitherFuture<A::ListenerUpgrade, B::ListenerUpgrade>;
     type Dial = EitherFuture<A::Dial, B::Dial>;
 
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<TransportEvent<Self>> {
+        match self.project() {
+            EitherTransportProj::Left(a) => match a.poll(cx) {
+                Poll::Pending => Poll::Pending,
+                Poll::Ready(event) => Poll::Ready(event.map(EitherFuture::First, EitherError::A)),
+            },
+            EitherTransportProj::Right(b) => match b.poll(cx) {
+                Poll::Pending => Poll::Pending,
+                Poll::Ready(event) => Poll::Ready(event.map(EitherFuture::Second, EitherError::B)),
+            },
+        }
+    }
+
     fn listen_on(
         &mut self,
+        id: ListenerId,
         addr: Multiaddr,
-    ) -> Result<Self::Listener, TransportError<Self::Error>> {
+    ) -> Result<(), TransportError<Self::Error>> {
```

Comment:
I think `listen_on` should return a `ListenerId` rather than expect one.

Now, I think this is a bit difficult because of how `Transport`s compose. The IDs should obviously be unique but they might not be if we compose `Transport`s together and sometimes return an ID from an inner transport and sometimes our own.

What if we "namespace" `ListenerId`s? We could extend `ListenerId` with a `TypeId` of the `Transport` they are tracking. E.g. each `Transport` would be encouraged to construct a their "current" `ListenerId` with `ListenerId::new::<Self>()`.

That would assume that each kind of `Transport` is only composed once. I think that is generally true although not impossible to violate. On the whole, I think it is a fair assumption to make though:

To users of `Transport` we can guarantee that the `ListenerId` returned within `TransportEvent`s unambiguously maps to their call to `listen_on` _as long as_ they compose `Transport`s correctly.

Reply:
Not sure if it's worth the extra complexity. I agree that it would be a cleaner API. But apart from that, what are the benefits of having the `ListenerId` generated by the transport?

Reply:
I think it is easier for clients to use if we don't require additional state tracking outside of a `Transport`. I am coming from the perspective of using these APIs _without_ `libp2p-swarm` and would rather have the `Transport`s itself take care of this state tracking.

The advantage is that if the `Transport` takes care of the state tracking, it is optional to use for the client (could just ignore the return value) whereas if it is a parameter, I HAVE to pass one to make the API work!

Additionally, a client may get the parameters wrong, hence the API is overall more complex to use.

Reply:
Good point. I only considered the `Transport` implementer's PoV and did not think about users that directly use transports without a swarm.

Reply:
More generally, having clients decide on identifiers is really useful if idempotency is required. If not, I'd default to having the "invoked system" (for lack of a better word) generate the identifier.

HTTP POST does this for example or databases and tables with primary keys.
I am guessing that one of the design choices for TCP's sequence number being client-generated is because it needs to replay packets in case they are not ACK'd: Idempotency.

Reply:
Resolving here since this discussion moved to another comment https://github.com/libp2p/rust-libp2p/pull/2652#discussion_r884311198. We agree that the inner Transport should generate the `ListenerId`, now it's more a questions of how the `ListenerId` API should look like.